src/internal/operators/mergeScan.ts

```diff
@@ -43,96 +44,78 @@ import { SimpleInnerSubscriber, SimpleOuterSubscriber, innerSubscribe } from '..
+      let index = 0;
+      // Whether or not the outer source has completed.
+      let isComplete = false;
 
-  protected _complete(): void {
-    this.hasCompleted = true;
-    if (this.active === 0 && this.buffer.length === 0) {
-      if (this.hasValue === false) {
-        this.destination.next(this.acc);
-      }
-      this.destination.complete();
-    }
-    this.unsubscribe();
-  }
+      /**
+       * Checks to see if we can complete our result or not.
+       */
+      const checkComplete = () => {
+        // If the outer has completed, and nothing is left in the buffer,
+        // and we don't have any active inner subscriptions, then we can
+        // Emit the state and complete.
+        if (isComplete && !buffer.length && !active) {
+          // TODO: This seems like it might result in a double emission, perhaps bad behavior?
+          // maybe we should change this in an upcoming major?
+          !hasState && subscriber.next(state);
+          subscriber.complete();
+        }
+      };
 
-  notifyNext(innerValue: R): void {
-    const { destination } = this;
-    this.acc = innerValue;
-    this.hasValue = true;
-    destination.next(innerValue);
-  }
+      const nextSourceValue = (value: T) => {
+        // If we're under our concurrency limit, go ahead and
+        // call the accumulator and subscribe to the result.
+        if (active < concurrent) {
+          active++;
+          from(accumulator(state!, value, index++)).subscribe(
+            new OperatorSubscriber(
+              subscriber,
+              (innerValue) => {
+                hasState = true;
+                // Intentially terse. Set the state, then emit it.
+                subscriber.next((state = innerValue));
+              },
+              undefined,
+              () => {
+                // The inner completed, decrement the number of actives.
+                active--;
+                // If we have anything in the buffer, process it, otherwise check to see if we can complete.
+                buffer.length ? nextSourceValue(buffer.shift()!) : checkComplete();
+              }
+            )
+          );
+        } else {
+          // We're over our concurrency limit, push it onto the buffer to be
+          // process later when one of our inners completes.
+          buffer.push(value);
+        }
+      };
 
-  notifyComplete(): void {
-    const buffer = this.buffer;
-    this.active--;
-    if (buffer.length > 0) {
-      this._next(buffer.shift());
-    } else if (this.active === 0 && this.hasCompleted) {
-      if (this.hasValue === false) {
-        this.destination.next(this.acc);
-      }
-      this.destination.complete();
-    }
-  }
+      source.subscribe(
+        new OperatorSubscriber(subscriber, nextSourceValue, undefined, () => {
+          // Outer completed, make a note of it, and check to see if we can complete everything.
+          isComplete = true;
+          checkComplete();
+        })
```

Comment:
question: This operator has a concurrency limit and buffer, too. Should it ensure the buffer is nulled/emptied upon explicit unsubscription? Should there be a general policy on this?

Reply:
>  Should there be a general policy/practice for this?

Yeah, I'm trying to standardize that in these new commits.